### DISCUSSION TITLE
Trying to access with query string parameter. Got: "Jwks doesn't have key to match kid or alg from Jwt"

### DISCUSSION THREAD

**Discussion by lorenzobassi96** [istio/istio#49244] [Q&A]
I am working on Istio 1.16.

In my environment, I have:

- A Kubernetes cluster
- Keycloak IDP (external to the cluster)
- oauth2proxy installed in the Kubernetes cluster.

I want to base my Authorization on the content of the query parameter.
`curl -k https://x.x.x.x/somepath/?access_token=$TOKEN`

Reading the documentation, I need to use `fromParam` filder in the RequestAuthentication

This is my RequestAuthentication:
```
apiVersion: security.istio.io/v1
kind: RequestAuthentication
metadata:
  name: xran-idp-authn
spec:
  jwtRules:
  - fromParams:
    - access_token
    forwardOriginalToken: false
    issuer: https://registry.local/auth/realms/jma
    jwksUri: https://registry.local/auth/realms/jma/protocol/openid-connect/certs
  selector: {}
```

When I try to get the service, 
`curl -k https://X.X.X.X/somepath/?access_token=$TOKEN`

I have the following:
`Jwks doesn't have key to match kid or alg from Jwt`

Looking at the logs in the istio sidecar, I see this:
```
2024-02-07T07:36:03.720238Z     debug   envoy http external/envoy/source/common/http/conn_manager_impl.cc:1680 [C21][S4126602580812274902] encoding headers via codec (end_stream=false):
':status', '401'
'www-authenticate', 'Bearer realm="https://X.X.X.X/somepath/?access_token=eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJQYTNub3lwY29aczl3RHNMYmxXWEJBN3lnSzd0RnhXa0RjYVlnaFF5QU9FIn0.eyJleHAiOjE3MDcyOTEzNzUsImlhdCI6MTcwNzI5MTA3NSwianRpIjoiNjVkZTM2ZmItMzgyNy00ZmZjLThlYzctMDM5NzNkNjczMDdjIiwiaXNzIjoiaHR0c", error="invalid_token"'
```

Can someone helm me in understanding the cause of this?
I did not find anything usefull for fixing this issue.

Thanks in advance

**Comment by yogeshgadge** [comment]
I have the same error - - any help appreciated

**Comment by yogeshgadge** [comment]
For me this error goes away if I specify `jwks` instead of the `jwksUri` but I should not be hard coding this.

**Comment by dangerouspython** [comment]
Hi, I recently came across this issue while trying to deploy a different environment are bringing down the previous one and I am using istio 1.24.2. I am actually following the Book *"Microservices with Spring Boot and Spring Cloud" 2nd edn. (Packt Publishing) by Magnus Larsson*. This issues is mentioned by the author on pg. 578 and here's it what he says and I am quoting verbatim (replace the words 'development' and 'production' by 'old' and 'new' respectively):

 This error is caused by the Istio daemon `istiod` caching the JWKS from the auth server in the development environment. The auth server in the production environment will have a new JWKS, but the same identity to `istiod`, so it tries to use the old JWKS, causing this failure. Istio caches a JWKS for 20 minutes, so you can simply take a long coffee break and the issue will be gone when you are back. Or you can empty the cache by restarting `istiod` a few times (2 or 3 times should be sufficient). Run the following command to restart the `istiod` Pod:
`
> kubectl -n istio-system delete pod -l app=istiod
`
Although this thread is quite old, I am just putting it out there for someone else's benefit 

(***NOTE***: the author uses v1.9.3, but i faced this issue with v1.24.2 - so, i guess the reasoning still applies)

**Comment by prasadzende** [comment]
Clearing browser cache and cookies resolved my issue.
